In [1]:
import pandas as pd
import numpy as np

from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split

from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LinearRegression

from sklearn.metrics import accuracy_score, confusion_matrix

Part A: Dataset Understanding

Q1.Load the dataset and display the first five records.

In [2]:
df = pd.read_csv("Dataset 2.csv")

df.head()

,UserID,Age,Gender,SubscriptionType,WatchHoursPerWeek,DevicesUsed,FavoriteGenre,AdClicks,MonthlySpend,SubscriptionRenewed
0,1001,22,Female,Basic,23,5,Comedy,13,353,No
1,1002,55,Male,Basic,9,4,Drama,14,317,Yes
2,1003,49,Male,Basic,8,3,Comedy,16,309,No
3,1004,39,Female,Premium,19,5,Drama,45,833,Yes
4,1005,38,Female,Premium,23,5,Sci-Fi,24,804,Yes


Q2.Determine the number of rows and columns in the dataset.

In [3]:
# Number of rows and columns
print("Rows and Columns:", df.shape)

Rows and Columns: (750, 10)


Q3. Display all column names.

In [4]:
print(df.columns)

Index(['UserID', 'Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek',
       'DevicesUsed', 'FavoriteGenre', 'AdClicks', 'MonthlySpend',
       'SubscriptionRenewed'],
      dtype='object')


Q4. Identify numerical and categorical features.

In [5]:
print(df.dtypes)

UserID                  int64
Age                     int64
Gender                 object
SubscriptionType       object
WatchHoursPerWeek       int64
DevicesUsed             int64
FavoriteGenre          object
AdClicks                int64
MonthlySpend            int64
SubscriptionRenewed    object
dtype: object


In [6]:
print("Q4 - Numerical Features:")
print(df.select_dtypes(include='number').columns.tolist())

print("\nQ4 - Categorical Features:")
print(df.select_dtypes(include='object').columns.tolist())

Q4 - Numerical Features:
['UserID', 'Age', 'WatchHoursPerWeek', 'DevicesUsed', 'AdClicks', 'MonthlySpend']

Q4 - Categorical Features:
['Gender', 'SubscriptionType', 'FavoriteGenre', 'SubscriptionRenewed']


 Check whether the dataset contains missing values.

In [7]:
print(df.isnull().sum())

UserID                 0
Age                    0
Gender                 0
SubscriptionType       0
WatchHoursPerWeek      0
DevicesUsed            0
FavoriteGenre          0
AdClicks               0
MonthlySpend           0
SubscriptionRenewed    0
dtype: int64


Part B: Exploratory Data Analysis

Q6. Calculate the average age of users.

In [8]:
print("Q6 - Average Age of Users:")
print(df["Age"].mean())

Q6 - Average Age of Users:
41.824


Q7. Determine the average watch hours per week.

In [9]:
print("Q7 - Average Watch Hours Per Week:")
print(df["WatchHoursPerWeek"].mean())

Q7 - Average Watch Hours Per Week:
14.236


Q8. Find the average monthly spending of users.

In [11]:
print("Q8 - Average Monthly Spending:")
print(df["MonthlySpend"].mean())

Q8 - Average Monthly Spending:
689.9053333333334


Q9. Count the number of users in each subscription category.

In [12]:
print("Q9 - Users in each Subscription Category:")
print(df["SubscriptionType"].value_counts())

Q9 - Users in each Subscription Category:
SubscriptionType
Basic      342
Premium    279
VIP        129
Name: count, dtype: int64


Q10. Determine the percentage of users who renewed their subscriptions.

In [13]:
print("Q10 - Percentage of users who renewed:")
renewed = (df["SubscriptionRenewed"]=="Yes").mean()*100
print(renewed)

Q10 - Percentage of users who renewed:
46.266666666666666


Part C: Data Preparation

Q11. Convert categorical features into numerical form.

In [14]:
print("Q11 - Converting categorical features to numerical:")
encoder = LabelEncoder()
df["Gender"] = encoder.fit_transform(df["Gender"])
df["SubscriptionType"] = encoder.fit_transform(df["SubscriptionType"])
df["FavoriteGenre"] = encoder.fit_transform(df["FavoriteGenre"])
df["SubscriptionRenewed"] = encoder.fit_transform(df["SubscriptionRenewed"])
print(df.head())

Q11 - Converting categorical features to numerical:
   UserID  Age  Gender  SubscriptionType  WatchHoursPerWeek  DevicesUsed  \
0    1001   22       0                 0                 23            5   
1    1002   55       1                 0                  9            4   
2    1003   49       1                 0                  8            3   
3    1004   39       0                 1                 19            5   
4    1005   38       0                 1                 23            5   

   FavoriteGenre  AdClicks  MonthlySpend  SubscriptionRenewed  
0              1        13           353                    0  
1              2        14           317                    1  
2              1        16           309                    0  
3              2        45           833                    1  
4              5        24           804                    1  


Q12. Define the feature set (X) and target variable (y) for subscription renewal prediction.

In [15]:
print("Q12 - Feature set and Target variable:")
X = df.drop(columns=["UserID","SubscriptionRenewed"])
y = df["SubscriptionRenewed"]
print("Features (X):", list(X.columns))
print("Target (y): SubscriptionRenewed")

Q12 - Feature set and Target variable:
Features (X): ['Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek', 'DevicesUsed', 'FavoriteGenre', 'AdClicks', 'MonthlySpend']
Target (y): SubscriptionRenewed


Q13. Split the dataset into training and testing sets.

In [16]:
print("Q13 - Splitting into Training and Testing sets:")
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print("Training samples:", X_train.shape[0])
print("Testing samples:", X_test.shape[0])

Q13 - Splitting into Training and Testing sets:
Training samples: 600
Testing samples: 150


Part D: Decision Tree Classification

Q14. Train a Decision Tree model to predict whether a user will renew their subscription.



In [18]:
dt = DecisionTreeClassifier(random_state=42)

dt.fit(X_train, y_train)

DecisionTreeClassifier(random_state=42)

Q15. Evaluate the model using accuracy.

In [19]:
print("Q15 - Decision Tree Accuracy:")
prediction = dt.predict(X_test)
accuracy = accuracy_score(y_test, prediction)
print("Accuracy:", accuracy)

Q15 - Decision Tree Accuracy:
Accuracy: 0.56


Q16. Generate and interpret the confusion matrix.

In [20]:
print("Q16 - Confusion Matrix:")
cm = confusion_matrix(y_test, prediction)
print(cm)
print("\nInterpretation:")
print("True Negatives  (TN) - Correctly predicted NOT renewing:", cm[0][0])
print("False Positives (FP) - Wrongly predicted as renewing:", cm[0][1])
print("False Negatives (FN) - Wrongly predicted as NOT renewing:", cm[1][0])
print("True Positives  (TP) - Correctly predicted as renewing:", cm[1][1])

Q16 - Confusion Matrix:
[[48 34]
 [32 36]]

Interpretation:
True Negatives  (TN) - Correctly predicted NOT renewing: 48
False Positives (FP) - Wrongly predicted as renewing: 34
False Negatives (FN) - Wrongly predicted as NOT renewing: 32
True Positives  (TP) - Correctly predicted as renewing: 36


In [ ]:
Part E: K-Nearest Neighbors (KNN)

Q17. Train a KNN classifier with K = 5.

In [21]:
print("Q17 - Training KNN Classifier with K=5:")
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)
print("KNN Model trained successfully!")

Q17 - Training KNN Classifier with K=5:
KNN Model trained successfully!


Q18. Compare the accuracy of KNN with the Decision Tree model.

In [23]:
print("Q18 - Comparing KNN vs Decision Tree:")
knn_prediction = knn.predict(X_test)
knn_accuracy = accuracy_score(y_test, knn_prediction)
print("KNN Accuracy:", knn_accuracy)
print("Decision Tree Accuracy:", accuracy)


Q18 - Comparing KNN vs Decision Tree:
KNN Accuracy: 0.6333333333333333
Decision Tree Accuracy: 0.56


Part F: Linear Regression

Q19. Train a Linear Regression model to predict monthly spending.

In [24]:
print("Q19 - Training Linear Regression to predict Monthly Spending:")
X_reg = df.drop(columns=["UserID", "MonthlySpend"])
y_reg = df["MonthlySpend"]
X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(X_reg, y_reg, test_size=0.2, random_state=42)
lr = LinearRegression()
lr.fit(X_train_reg, y_train_reg)
print("Linear Regression Model trained successfully!")

Q19 - Training Linear Regression to predict Monthly Spending:
Linear Regression Model trained successfully!


Q20. Predict the monthly spending for a new user and interpret the result.


In [27]:
new_user = [22, 0, 0, 15, 2, 1, 13, 0]
prediction = lr.predict([new_user])
print("Predicted Monthly Spend:", prediction[0])

Predicted Monthly Spend: 349.1532651068479


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


**Business Reflections**

**Q1. Which factors influence subscription renewal most?**

Answer:- Likely: WatchHoursPerWeek MonthlySpend SubscriptionType DevicesUsed

These factors reflect user engagement.

**Q2. Why is subscription renewal a classification problem?**

Answer:- Because the output is:

Yes or No

Fixed categories.

**Q3. Why is monthly spending a regression problem?**

Answer:- Because spending can be any numerical value such as:

₹300 ₹500 ₹800 ₹1200

Continuous numbers.

**Q4. Which algorithm performed better?** **bold text**

Answer:-

KNN

KNN Accuracy = 63.33%% Decision Tree Accuracy = 56%

Q5. **How can Netflix use these predictions?**

Answer:-

Identify users likely to leave. Offer discounts. Send personalized recommendations. Improve customer retention. Increase revenue.